### PR TITLE
NOT FOR SUBMISSION (yet) - prototype of enum casing for C#

### PR DESCRIPTION
**Pull request by jskeet**
The implements prefix removal and PascalCasing, so in an enum Color,
a value of COLOR_BLUE would end up as Blue in the generated code.

This initial code is unconditional; in reality we may want it controlled by a flag.
We'd also definitely want to improve the performance of JsonFormatter.GetOriginalEnumValueName, which uses reflection and a new attribute to determine the original name.

Divided into four commits:
- C++ code changes
- C# library changes
- C# test changes (just changing how we refer to the enum values)
- C# generated code changes

Josh: apologies for any C++ faux-pas. We can definitely improve the standard of that code before submission :)

Related issues: #1334 and #1330.

### CODE REVIEW

**Comment by haberman** [comment]
ok to test

File: src/google/protobuf/compiler/csharp/csharp_helpers.cc [hunk 1]
```diff
@@ -240,6 +239,121 @@ std::string GetPropertyName(const FieldDescriptor* descriptor) {
...
+// None - first char of input    Alphanumeric              Upper
+// Non-letter (e.g. _ or 1)      Alphanumeric              Upper
+// Numeric                       Alphanumeric              Upper
+// Lower letter                  Alphanumeric              Same as current
+// Upper letter                  Alphanumeric              Lower
+std::string ShoutyToPascalCase(const std::string& input) {
+  string result;
+  // Simple way of implementing "always start with upper"
+  char previous = '_';
+  for (int i = 0; i < input.size(); i++) {
```
Comment (ivannaranjo) [inline src/google/protobuf/compiler/csharp/csharp_helpers.cc]: Use a range for here, clearer intent.

In general prefer using ++i when iterating, irrelevant performance with ints but it is more idiomatic.

File: src/google/protobuf/compiler/csharp/csharp_helpers.cc [hunk 2]
```diff
@@ -240,6 +239,121 @@ std::string GetPropertyName(const FieldDescriptor* descriptor) {
...
+  }
+  return result;
+}
+
+// Generate the name to use in a generated C# enum from the enum name, the enum value name, and the naming style.
+std::string GetEnumValueName(const std::string& enum_name, const std::string& enum_value_name, CSharpEnumNaming namingStyle) {
+  if (namingStyle == CSHARPENUMNAMING_PRESERVE) {
+    return enum_value_name;
+  }
+  std:string effective_value_name = enum_value_name;
```
Comment (ivannaranjo) [inline src/google/protobuf/compiler/csharp/csharp_helpers.cc]: Declare `effective_value_name` as a parameter by value in place of `enum_value_name`, avoid unnecessary copies.

Comment (ivannaranjo) [inline src/google/protobuf/compiler/csharp/csharp_helpers.cc]: Copy `enum_value_name` into `effective_value_name` in an else block instead, avoid this copy when `namingStyle` is `CSHARPENUMNAMING_PASCAL_TRIM_PREFIX`.

**Comment by jskeet** [comment]
(A better version of this was merged in #1401.)

**Pull request closed without merging**